src/picongpu/include/traits/attribute/GetChargeState.hpp

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright 2014 Rene Widera
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "simulation_defines.hpp"
+#include "traits/HasIdentifier.hpp"
+#include "algorithms/TypeCast.hpp"
+#include "particles/traits/GetProtonNumber.hpp"
+
+namespace picongpu
+{
+namespace traits
+{
+namespace attribute
+{
+namespace detail
+{
+
+/** Calculate the charge state of an atom / ion
+ *
+ * use attribute `boundElectrons` to calculate the charge state
+ */
+template<bool T_HasBoundElectrons>
+struct LoadChargeState
+{
+
+    template<typename T_Particle>
+    HDINLINE float_X operator()(const T_Particle& particle)
+    {
+        const uint32_t protonNumber = GetProtonNumber<T_Particle>::type::value;
+        return precisionCast<float_X>(protonNumber) - particle[boundElectrons_];
+    }
+};
+
+/**  Calculate charge state of an atom / ion
+ *
+ * This is the fallback implementation if no `boundElectrons` are available for a particle
+ */
+template<>
+struct LoadChargeState<false>
+{
+
+    template<typename T_Particle>
+    HDINLINE float_X operator()(const T_Particle& particle)
+    {
+        return 1.0;
```

Comment:
That is not what I'm doing. The thing is that as long as ions get cloned out of electrons, I cannot start a simulation with neutral particles. If that particle has `boundElectrons` then I can calculate a `chargeState`. If it doesn't have `boundElectrons` it is most likely an electron and thus singly charged. I could also put an error message in there. "It is no use to ask for a charge state of particles that have no bound electrons because then they are probably elementary particles and their charge does not change. But if you still want to: default is 1 because ```charge = CHARGE_CONST * chargeState```." The whole thing is just too preliminary. 
What should happen if you try to call `getChargeState` on a particle without differing charge states?